### DISCUSSION TITLE
MongoDB's ObjectId (BSON) on the browser

### DISCUSSION THREAD

**Discussion by italodeandra** [vercel/next.js#31776] [Help]
Hi team,

I'm currently working on a Next.js project where I use MongoDB as a the database. To decrease the bundle size, I removed MongoDB from the browser with webpack but I needed the frontend to at least have the `ObjectId`, so I started using the `bson` package for that.

The problem is: `ObjectId` needs some kind of crypto, and since the browser doesn't have that I think that Next.js installs a "browserfy" version of it, making the bundle size skyrocket. Here's my project bundle analysis and as you can see, the dependencies packages meant for crypto together is bigger than Material UI. The crypto dependencies are those called "bn" and "elliptic".

![image](https://user-images.githubusercontent.com/19225266/143305272-20403f31-070e-42ae-9732-965a87ea8fc5.png)

The thing I want to discuss here is: **what should I do?** I want to keep the ObjectId on the browser but I don't want to pay the price of 200kb of gzipped bundle size (currently sitting on 500kb gzipped).

Is the best option (or maybe only) just removing the ObjectId entirely from the browser and just using its hex string?

**Comment by wwwdepot** [comment]
> I want to keep the ObjectId on the browser
Why?

**Reply by italodeandra** [reply]
1. Consistency. I'm currently using `superjson` to serialize/deserialize my data and have them being the exact same of both browser and server.
2. In some specific situations I create new IDs from the frontend application instead of letting the Mongo create them. Not possible to let the backend create them, otherwise I'll lose a lot of speed since I'll need to keep going to the backend several times. This way I can batch them and avoid two of them having the same ID. Any other ID generation solution is no better than just using bson's ObjectId so I decided to just keep using it.
3. OCD.

**Reply by wwwdepot** [reply]
Ok. Probably want to lazy load the library when needed. 

```jsx
import { useState } from "react";

export default function Page() {
  const [id, setId] = useState("");

  async function generateId() {
    const ObjectId = (await import("bson")).ObjectID;
    setId(new ObjectId().toString());
  }
  return (
    <div>
      <button onClick={generateId}>Generate ID</button>
      <pre>ID: {id}</pre>
    </div>
  );
}
```

**Reply by italodeandra** [reply]
That's a nice idea. But I actually also use the ObjectId when deserializing the data from `getServerSideProps` (so I can compare them with the newly created ones) then this will be lazy loaded when the page loads anyway. But thanks!

I think I'll probably just keep the extra 200kb. I'll leave this discussion open because maybe some new ideas might come up.
